source/src/cip/cipqos.c

```diff
@@ -25,6 +25,24 @@ CipUsint dscp_low = 31; /**< DSCP value for CIP transport class 0/1 low priority
 CipUsint dscp_explicit = 27; /**< DSCP value for CIP explicit messages (transport class 2/3 and UCMM)
                                 and all other EtherNet/IP encapsulation messages */
 
+/** @brief Hidden copy of data that need to be frozen on boot-up
+ *
+ *  The QoS DSCP values can be changed from the EIP network but the changes should come
+ *  into effect only after a restart. Therefore we take a copy here when the network is
+ *  initialized in NetworkHandlerInitialize() and CipQosGetDscpPriority() is called the
+ *  first time.
+ */
+static bool s_is_frozen = false;
+static struct {
+  CipUsint dscp_event;
+  CipUsint dscp_general;
+  CipUsint dscp_urgent;
+  CipUsint dscp_scheduled;
+  CipUsint dscp_high;
+  CipUsint dscp_low;
+  CipUsint dscp_explicit;
+} s_frozen;
```

Comment:
give struct a type name, and a better var name

Reply:
My reason to give the variable neither a struct name nor a type definition was that this variable is static and there should only be a single instance of it (as long as the code doesn't support multiple EIP ports). Not declaring a structure type nor giving a type definition should deter the user from creating another instance.

Now having looked at the code again perhaps the "bool s_is_frozen" variable could be made to a member of the structure (s_frozen) something like "bool valid;".
And in turn the struct could be named "s_dscp_bootup_state".

Reply:
I thought so, but perhaps someone out there already has an OpENer version with multiple ports (not sharing it back), so this would always put additional effort on such possible developers. Putting that into the C file (and possible marking it as Private in a comment) should suffice

Reply:
I think doing so as you proposed with the typedef should be fine.

When I was authoring the original patch I also pondered if the saving of the bootup state of the DSCP values should be made explicit and a dedicated function like CipQosLockDscpBootupState() should be created. That should then be called from NetworkHandlerInitialize() before the first call to CipQosGetDscpPriority() is made.

But that still should be called from CipQosGetDscpPriority() if s_dscp_bootup_state.valid is not set.

Only as an afterthought ...

Reply:
I did it right now, similar to your proposal, but dropped the valid flag entirely in favor of a dedicated update function, which is called by the Reset function. The initial set of default values is right now a set of defines, which act as initialization values for the attributes and the currently used set of DSCP values. SetAttribute then writes into the instance attributes, but only a reset triggers the transition to the working set, via the update function.

Unfortunately my Linux VM currently behaves a bit strange, so I cannot fully test it with the CT tool.
But I will push my changes for further review.